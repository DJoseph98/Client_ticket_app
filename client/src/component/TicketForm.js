import { React, useState } from 'react';
import TextField from '@material-ui/core/TextField';
import { makeStyles } from '@material-ui/core/styles';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import FormHelperText from "@material-ui/core/FormHelperText";
import Button from '@material-ui/core/Button';
import Box from '@material-ui/core/Box';

const useStyles = makeStyles((theme) => ({
    root: {
        display: 'flex',
        flexWrap: 'wrap',
    },
    textField: {
        marginLeft: theme.spacing(1),
        marginRight: theme.spacing(1),
        width: '25ch',
    },
}));

const TicketForm = (props) => {
    const classes = useStyles();
    const [hasError, sethasError] = useState(false);
    const [hasErrorEmail, setHasErrorEmail] = useState(false);
    const [formData, setFormData] = useState({
        title: props.ticket ? props.ticket.title : '',
        email: props.ticket ? props.ticket.email : '',
        problemDescription: props.ticket ? props.ticket.problemDescription : '',
        ticketLvlPriorId: props.ticket ? props.ticket.Ticket_Level_Priority.id : 1,
        ticketStatusId: props.ticket ? props.ticket.Ticket_Status.id : 1,
    })
    const setChangedState = (stateName) => {
        setFormData(formData => ({ ...formData, ...stateName }));
    }
    const titleChange = (e) => {
        setChangedState({ title: e.target.value });
    }
    const emailChange = (e) => {
        setChangedState({ email: e.target.value });
    }
    const pbDescriptionChange = (e) => {
        setChangedState({ problemDescription: e.target.value });
    }
    const ticketStatusChange = (e) => {
        setChangedState({ ticketStatusId: e.target.value });
    }
    const ticketLvlPriorChange = (e) => {
        setChangedState({ ticketLvlPriorId: e.target.value });
    }
    const handleFormSubmit = (e) => {
        e.preventDefault();
        sethasError(false);
        setHasErrorEmail(false);
        if (props.ticket && formData.ticketStatusId === 1) {
            sethasError(true);
            return false;
        } else if (formData.email.match(/^(([^<>()[\]\.,;:\s@\"]+(\.[^<>()[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/) === null) {
            setHasErrorEmail(true);
            return false;
        }

        let newFormData = new FormData(e.target);
        props.onSubmit(Object.fromEntries(newFormData));
    };
    return (
        <div>
            <form className={classes.root} onSubmit={handleFormSubmit}>
                <FormControl style={{ margin: 8 }} fullWidth>
                    <TextField
                        labelId="label_title"
                        name="title"
                        label="Title"
                        type="text"
                        onChange={titleChange}
                        variant="outlined"
                        inputProps={{
                            maxLength: 50
                        }}
                        helperText={`${formData.title.length}/50`}
                        defaultValue={formData.title}
                        value={formData.title}
                        placeholder="Title"
                        fullWidth
                        required />
                </FormControl>
                <FormControl style={{ margin: 8 }} fullWidth>
                    <TextField
                        name="email"
                        label="Email"
                        type="email"
                        onChange={emailChange}
                        defaultValue={formData.email}
                        placeholder="Email"
                        variant="outlined"
                        error={hasErrorEmail}
                        required
                    />
                    {hasErrorEmail && <FormHelperText>Incorrect email format</FormHelperText>}
                </FormControl>
                <TextField
                    type="text"
                    name="problemDescription"
                    style={{ margin: 8 }}
                    inputProps={{
                        maxLength: 700
                    }}
                    helperText={`${formData.problemDescription.length}/700`}
                    label="Description"
                    multiline
                    rows={10}
                    variant="outlined"
                    defaultValue={formData.problemDescription}
                    onChange={pbDescriptionChange}
                    fullWidth
                    required
                />
                <FormControl variant="outlined" fullWidth style={{ margin: 8 }}>
                    <InputLabel id="label_priority">Priority</InputLabel>
                    <Select
                        labelId="label_priority"
                        value={formData.ticketLvlPriorId}
                        onChange={ticketLvlPriorChange}
                        required
                        label="Priority"
                        name="ticketLvlPriorId"
                    >
                        <MenuItem value={1}>Blocker</MenuItem>
                        <MenuItem value={2}>Critical</MenuItem>
                        <MenuItem value={3}>High</MenuItem>
                        <MenuItem value={4}>Normal</MenuItem>
                        <MenuItem value={5}>Low</MenuItem>
                    </Select>
                </FormControl>

                {props.ticket && props.ticket.Ticket_Activity.activity !== 'pending' &&
                    <FormControl variant="outlined" fullWidth style={{ margin: 8 }} error={hasError}>
                        <InputLabel id="label_status">Status</InputLabel>
                        <Select
                            labelId="label_status"
                            value={formData.ticketStatusId}
                            onChange={ticketStatusChange}
                            required
                            label="Status"
                            name="ticketStatusId"
                        >
                            <MenuItem value={2}>In progress</MenuItem>
                            <MenuItem value={3}>Closed</MenuItem>
                        </Select>
                        {hasError && <FormHelperText>This is required!</FormHelperText>}
                    </FormControl>
                }

                <div style={{ width: '100%' }}>
                    <Box display="flex" justifyContent="flex-end" style={{ margin: 8 }}>
                        {props.ticket
                            ? <Button id="upt" type="submit" variant="contained" size="large" color="primary">Update Ticket</Button>
                            : <Button type="submit" variant="contained" size="large" color="primary">Create Ticket</Button>
                        }
                    </Box>
                </div>
            </form>
        </div>
    );
};

export default TicketForm;