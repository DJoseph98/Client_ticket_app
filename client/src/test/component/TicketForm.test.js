import React from 'react';
import { shallow } from 'enzyme';
import TicketForm from '../../component/TicketForm';
import ticketFixtures from '../fixtures/tickets';

const entries = jest.fn()
const append = jest.fn()
global.FormData = () => ({ entries, append })

describe('Ticket Form page', () => {
    let wrapper;
    beforeAll(() => {  // This is Mocha; in Jest, use beforeAll
        wrapper = shallow(<TicketForm />);
    });

    it('should be form page without data', () => {
        expect(wrapper).toMatchSnapshot();
    })
    it('should be form page with data', () => {
        const wrapper = shallow(<TicketForm ticket={ticketFixtures[0]} />);
        expect(wrapper).toMatchSnapshot();
    })
    it('should test title change', () => {
        const value = 'hello';
        wrapper.find({ name: "title" }).simulate('change', {
            target: { value }
        });
        expect(wrapper.find({ name: "title" }).props().defaultValue).toBe(value);
    });
    it('should test email change', () => {
        const value = 'dza@dzea';
        wrapper.find({ name: "email" }).simulate('change', {
            target: { value }
        });
        expect(wrapper.find({ name: "email" }).props().defaultValue).toBe(value);
    });
    it('should test problemDescription change', () => {
        const value = 'probleme test ';
        wrapper.find({ name: "problemDescription" }).simulate('change', {
            target: { value }
        });
        expect(wrapper.find({ name: "problemDescription" }).props().defaultValue).toBe(value);
    });
    it('should test ticketLvlPriorId change', () => {
        const value = 3;
        wrapper.find({ name: "ticketLvlPriorId" }).simulate('change', {
            target: { value }
        });
        expect(wrapper.find({ name: "ticketLvlPriorId" }).props().value).toBe(value);
    });
    it('should test ticketStatusId change', () => {
        const value = 3;
        const wrapper = shallow(<TicketForm ticket={ticketFixtures[0]} />);
        wrapper.find({ name: "ticketStatusId" }).simulate('change', {
            target: { value }
        });
        expect(wrapper.find({ name: "ticketStatusId" }).props().value).toBe(value);
    });
    it('should test submit value from form', () => {
        const onSubmitSpy = jest.fn();
        const wrapper = shallow(<TicketForm ticket={ticketFixtures[0]} onSubmit={onSubmitSpy} />);

        wrapper.find('form').simulate('submit', {
            preventDefault: () => {
            }
        });
        
         expect(onSubmitSpy).toHaveBeenLastCalledWith({
             title: ticketFixtures[0].title,
             email: ticketFixtures[0].email,
             problemDescription: ticketFixtures[0].problemDescription,
             ticketLvlPriorId: ticketFixtures[0].Ticket_Level_Priority.id.toString(),
             ticketStatusId: ticketFixtures[0].Ticket_Status.id.toString()
         });
    });
});