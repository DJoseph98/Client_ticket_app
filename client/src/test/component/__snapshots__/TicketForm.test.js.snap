// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ticket Form page should be form page with data 1`] = `ShallowWrapper {}`;

exports[`Ticket Form page should be form page without data 1`] = `ShallowWrapper {}`;
